### PR TITLE
Combine resolvers

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary
Combining the resolver-classes into a resolver-module will reduce code duplicity (combined 2 methods) - and help add this functionality for other components in the future.

support the issue: #1335

## Quick changelog

- move strategy and hyperopt resolvers to resolver package
- convert from os.path to the more modern pathlib

### CODE REVIEW

**Comment by mishaker** [comment]
Nice first step ! bravo !

**Comment by mishaker** [comment]
This is exactly what we need for exchange and OrderHandler !

File: freqtrade/resolvers/iresolver.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import logging
+from pathlib import Path
+from typing import Optional, Dict, Type, Any
+
+logger = logging.getLogger(__name__)
+
+
+class IResolver(object):
+    """
+    This class contains all the logic to load custom hyperopt class
```
Comment (mishaker) [inline freqtrade/resolvers/iresolver.py]: copy paste? it is not just for hyperopt.

File: freqtrade/resolvers/iresolver.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    """
+
+    def __init__(self, object_type, config: Optional[Dict] = None) -> None:
```
Comment (mishaker) [inline freqtrade/resolvers/iresolver.py]: what good does it do object_type here ?

File: freqtrade/resolvers/iresolver.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        # Generate spec based on absolute path
+        spec = importlib.util.spec_from_file_location('unknown', str(module_path))
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)  # type: ignore # importlib does not use typehints
+
+        valid_objects_gen = (
+            obj for name, obj in inspect.getmembers(module, inspect.isclass)
+            if object_name == name and object_type in obj.__bases__
+        )
+        return next(valid_objects_gen, None)
```
Comment (mishaker) [inline freqtrade/resolvers/iresolver.py]: next(...) will return the next item from the iterator but the parent function name is _get_valid_objects. just thinking loud ...

File: freqtrade/resolvers/iresolver.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    @staticmethod
+    def _search_object(directory: Path, object_type, object_name: str,
+                       kwargs: dict = {}) -> Optional[Any]:
+        """
+        Search for the objectname in the given directory
+        :param directory: relative or absolute directory path
+        :return: object instance
+        """
+        logger.debug('Searching for %s %s in \'%s\'', object_type.__name__,  object_name, directory)
```
Comment (mishaker) [inline freqtrade/resolvers/iresolver.py]: two spaces after "__name__,  " flake made me obsessed :D

**Comment by mishaker** [comment]
The consistency check can be simply done by a list of all other "official" supported resolvers. otherwise generates a warning. If OrderHandler resolver implements on exchange limit stoploss it includes in consistency_check function: 
consistency_check += resolver(exchange).name == 'binance'
Syntax in my example is shit but you got the idea.

File: freqtrade/resolvers/iresolver.py (hunk 4 above)
Comment (xmatthias) [reply freqtrade/resolvers/iresolver.py]: i wonder why flake didn't complain about this ...

File: freqtrade/resolvers/iresolver.py (hunk 2 above)
Comment (xmatthias) [reply freqtrade/resolvers/iresolver.py]: good point, the whole __init__ is not needed, as we need to subclass this anyway

**Comment by xmatthias** [comment]
I disagree, the consistency-check cannot not be generalized and done in IResolver, but in each subclass if needed at loading time.
No consistency-check is needed for Strategy or Hyperopt.
It checks if the class loaded inherits from the correct parent class (IStrategy, IHyperopt) - but that's all IResolver does and should do. If more checks are needed to load the correct class, these need to be done in StrategyResolver or HyperoptResolver respectively.

The checks you're talking about are also not an issue of a loader / resolver - but configuration issues you're talking about should not be taken on at this level.
You always have to load one thing first, either exchange or strategy - so one doesn't know about the other necessarily (and they shouldn't, because the 2 things are not really related). It's the bot's excercise to do the checking - not the resolvers.

File: freqtrade/resolvers/iresolver.py (hunk 3 above)
Comment (xmatthias) [reply freqtrade/resolvers/iresolver.py]: you're right, the function name is misleading (i think a long time ago it did something else ...) 
I'll try to rename it to something more appropriate

**Comment by mishaker** [comment]
you are right, this is not a proper place. I was thinking about just a check_consistency abstract function to give e hint to the one who is implementing an OrderHandler or Strategy that it may or not be compatible. but not the right place. anyway I think you got the idea. I leave it to you cause I don't have a clear idea of how we can implement it for now ... see this one as the way I found currently: https://github.com/freqtrade/freqtrade/pull/1341/commits/519b1f00e2932d034afb76ee999f4273b8c7467a#diff-286b4739761aa953f7c682f8a87bcc81R72

**Pull request merged**